// typedoc-plugin-markdown/dist/app/renderer/overrides.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.render = exports.generateDocs = void 0;
const markdown_page_event_1 = require("../events/markdown-page-event");
const markdown_renderer_event_1 = require("../events/markdown-renderer-event");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const typedoc_1 = require("typedoc");
/**
 * Contains functionality to decouple HTML logic from the TypeDoc's {@link Renderer}.
 *
 * @todo Investigate ways to properly decouple HTML logic from the TypeDoc renderer.
 *
 * @module
 */
/**
 * Replacement of TypeDoc's {@link Application.generateDocs} method to decouple HTML logic.
 *
 */
async function generateDocs(project, out) {
    const start = Date.now();
    await this.renderer.render(project, out);
    if (this.logger.hasErrors()) {
        this.logger.error('Documentation could not be generated due to the errors above.');
    }
    else {
        this.logger.info(`Documentation generated at ${nicePath(out)}`);
        this.logger.verbose(`Markdown rendering took ${Date.now() - start}ms`);
    }
}
exports.generateDocs = generateDocs;
/**
 * Replacement of TypeDoc's {@link Renderer.render} method to decouple HTML logic.
 *
 * This is essentially a copy of the base method with a few tweaks.
 *
 * - Removes unnecessary async calls to load highlighters only required for html theme.
 * - Removes hooks logic that are jsx specific.
 * - Adds any logic specific to markdown rendering.
 */
async function render(project, outputDirectory) {
    this.renderStartTime = Date.now();
    if (this.cleanOutputDir) {
        try {
            fs.rmSync(outputDirectory, { recursive: true, force: true });
        }
        catch (error) {
            this.application.logger.warn('Could not empty the output directory.');
            return;
        }
    }
    try {
        fs.mkdirSync(outputDirectory, { recursive: true });
    }
    catch (error) {
        this.application.logger.error(`Could not create output directory ${outputDirectory}.`);
        return;
    }
    this.prepareTheme();
    const output = new markdown_renderer_event_1.MarkdownRendererEvent(markdown_renderer_event_1.MarkdownRendererEvent.BEGIN, outputDirectory, project);
    output.urls = this.theme.getUrls(project);
    output.navigation = this.theme.getNavigation(project);
    this.trigger(output);
    await Promise.all(this.preRenderAsyncJobs.map((job) => job(output)));
    this.preRenderAsyncJobs = [];
    if (!output.isDefaultPrevented) {
        this.application.logger.verbose(`There are ${output.urls?.length} pages to write.`);
        output.urls
            ?.filter((urlMapping) => urlMapping.model instanceof typedoc_1.ProjectReflection ||
            urlMapping.model instanceof typedoc_1.DeclarationReflection)
            .forEach(async (urlMapping) => {
            const [template, page] = output.createPageEvent(urlMapping);
            this.trigger(markdown_page_event_1.MarkdownPageEvent.BEGIN, page);
            if (page.isDefaultPrevented) {
                return false;
            }
            if (page.model instanceof typedoc_1.Reflection) {
                page.contents = this.theme.render(page, template);
            }
            else {
                throw new Error('Should be unreachable');
            }
            this.trigger(markdown_page_event_1.MarkdownPageEvent.END, page);
            if (page.isDefaultPrevented) {
                return false;
            }
            try {
                writeFileSync(page.filename, page.contents);
            }
            catch (error) {
                this.application.logger.error(`Could not write ${page.filename}`);
            }
        });
        await Promise.all(this.postRenderAsyncJobs.map((job) => job(output)));
        this.postRenderAsyncJobs = [];
        this.trigger(markdown_renderer_event_1.MarkdownRendererEvent.END, output);
    }
    this.theme = void 0;
}
exports.render = render;
/**
 * Some useful utility functions - essentially cherry picked from:
 *
 * - https://github.com/TypeStrong/typedoc/blob/master/src/lib/utils/fs.ts
 * - https://github.com/TypeStrong/typedoc/blob/master/src/lib/utils/path.ts
 */
/**
 *  Writes a file to disc.
 */
function writeFileSync(fileName, data) {
    fs.mkdirSync(path.dirname(normalizePath(fileName)), { recursive: true });
    fs.writeFileSync(normalizePath(fileName), data);
}
/**
 * Returns a readable path from an absolute path.
 */
function nicePath(absPath) {
    if (!path.isAbsolute(absPath))
        return absPath;
    const relativePath = path.relative(process.cwd(), absPath);
    if (relativePath.startsWith('..')) {
        return normalizePath(absPath);
    }
    return `./${normalizePath(relativePath)}`;
}
/**
 * Normalizes directory separators from a given path.
 */
function normalizePath(path) {
    return path.replace(/\\/g, '/');
}
