// typedoc-plugin-markdown/dist/theme/markdown-theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarkdownTheme = void 0;
const utils_1 = require("../libs/utils");
const options_1 = require("../options/index.js");
const typedoc_1 = require("typedoc");
const navigation_builder_1 = require("./core/navigation-builder");
const url_builder_1 = require("./core/url-builder");
const markdown_themecontext_1 = require("./markdown-themecontext");
/**
 * The main theme class for the plugin.
 *
 * The class controls how TypeDoc models are mapped to files and templates and extends TypeDoc's base Theme class.
 *
 * You would typically only be interested in overriding the the theme's render context instance.
 *
 * The API follows the implementation of [TypeDoc's custom theming](https://github.com/TypeStrong/typedoc/blob/master/internal-docs/custom-themes.md) with some minor adjustments.
 *
 * @usage
 *
 * ```ts
 * export function load(app) {
 *   app.renderer.defineTheme('customTheme', MyMarkdownTheme);
 * }
 *
 * class MyMarkdownTheme extends MarkdownTheme {
 *  ...
 * }
 * ```
 *
 * @category Theme
 */
class MarkdownTheme extends typedoc_1.Theme {
    /**
     * @ignore
     */
    constructor(renderer) {
        super(renderer);
        /**
         * @internal
         */
        this.readmeTemplate = (page) => {
            return this.getRenderContext(page).templates.readme();
        };
        /**
         * @internal
         */
        this.projectTemplate = (page) => {
            return this.getRenderContext(page).templates.project();
        };
        /**
         * @internal
         */
        this.reflectionTemplate = (page) => {
            return this.getRenderContext(page).templates.reflection();
        };
        this.textContentMappings = {
            ...options_1.TEXT_MAPPING_DEFAULTS,
            ...(this.application.options.getValue('textContentMappings') || {}),
        };
    }
    /**
     * Renders a template and page model to a string.
     *
     * @internal
     */
    render(page, template) {
        try {
            return (0, utils_1.formatMarkdown)(template(page));
        }
        catch (e) {
            console.log(e);
            this.application.logger.error(`Error rendering page ${page.url}.`);
            throw new Error(e);
        }
    }
    /**
     * Creates a new instance of the current theme context.
     *
     * This method can be overridden to provide an alternative theme context.
     */
    getRenderContext(page) {
        return new markdown_themecontext_1.MarkdownThemeContext(this, page, this.application.options);
    }
    /**
     * Maps the models of the given project to the desired output files.
     *
     * This method can be overriden to provide an alternative url structure.
     */
    getUrls(project) {
        return (0, url_builder_1.buildUrls)(this, project);
    }
    /**
     * Map the models of the given project to a navigation structure.
     *
     * This method can be overriden to provide an alternative navigation structure.
     */
    getNavigation(project) {
        return (0, navigation_builder_1.buildNavigation)(this, project);
    }
    /**
     * @internal
     */
    getTemplateMapping(kind, outputFileStrategy) {
        outputFileStrategy =
            outputFileStrategy ||
                this.application.options.getValue('outputFileStrategy');
        const directoryName = (reflectionKind) => {
            const pluralString = typedoc_1.ReflectionKind.pluralString(reflectionKind);
            return pluralString.replace(/[\s_-]+/g, '-').toLowerCase();
        };
        const membersWithOwnFile = this.application.options.getValue('membersWithOwnFile');
        const memberMapping = (template, kind) => {
            return {
                template,
                directory: directoryName(kind),
                kind: kind,
            };
        };
        const mappings = {
            [typedoc_1.ReflectionKind.Module]: {
                template: this.reflectionTemplate,
                directory: null,
                kind: typedoc_1.ReflectionKind.Module,
            },
            [typedoc_1.ReflectionKind.Namespace]: {
                template: this.reflectionTemplate,
                directory: directoryName(typedoc_1.ReflectionKind.Namespace),
                kind: typedoc_1.ReflectionKind.Namespace,
            },
        };
        if (outputFileStrategy === options_1.OutputFileStrategy.Members &&
            membersWithOwnFile?.includes(typedoc_1.ReflectionKind[typedoc_1.ReflectionKind.Class])) {
            mappings[typedoc_1.ReflectionKind.Class] = memberMapping(this.reflectionTemplate, typedoc_1.ReflectionKind.Class);
        }
        if (outputFileStrategy === options_1.OutputFileStrategy.Members &&
            membersWithOwnFile?.includes(typedoc_1.ReflectionKind[typedoc_1.ReflectionKind.Interface])) {
            mappings[typedoc_1.ReflectionKind.Interface] = memberMapping(this.reflectionTemplate, typedoc_1.ReflectionKind.Interface);
        }
        if (outputFileStrategy === options_1.OutputFileStrategy.Members &&
            membersWithOwnFile?.includes(typedoc_1.ReflectionKind[typedoc_1.ReflectionKind.Enum])) {
            mappings[typedoc_1.ReflectionKind.Enum] = memberMapping(this.reflectionTemplate, typedoc_1.ReflectionKind.Enum);
        }
        if (outputFileStrategy === options_1.OutputFileStrategy.Members &&
            membersWithOwnFile?.includes(typedoc_1.ReflectionKind[typedoc_1.ReflectionKind.Function])) {
            mappings[typedoc_1.ReflectionKind.Function] = memberMapping(this.reflectionTemplate, typedoc_1.ReflectionKind.Function);
        }
        if (outputFileStrategy === options_1.OutputFileStrategy.Members &&
            membersWithOwnFile?.includes(typedoc_1.ReflectionKind[typedoc_1.ReflectionKind.TypeAlias])) {
            mappings[typedoc_1.ReflectionKind.TypeAlias] = memberMapping(this.reflectionTemplate, typedoc_1.ReflectionKind.TypeAlias);
        }
        if (outputFileStrategy === options_1.OutputFileStrategy.Members &&
            membersWithOwnFile?.includes(typedoc_1.ReflectionKind[typedoc_1.ReflectionKind.Variable])) {
            mappings[typedoc_1.ReflectionKind.Variable] = memberMapping(this.reflectionTemplate, typedoc_1.ReflectionKind.Variable);
        }
        return mappings[kind];
    }
}
exports.MarkdownTheme = MarkdownTheme;
