// astro/dist/vite-plugin-astro-server/pipeline.js
import url from "node:url";
import { getInfoOutput } from "../cli/info/index.js";
import { ASTRO_VERSION, DEFAULT_404_COMPONENT } from "../core/constants.js";
import { enhanceViteSSRError } from "../core/errors/dev/index.js";
import { AggregateError, CSSError, MarkdownError } from "../core/errors/index.js";
import { Pipeline, loadRenderer } from "../core/render/index.js";
import { isPage, resolveIdToUrl, viteID } from "../core/util.js";
import { isServerLikeOutput } from "../prerender/utils.js";
import { PAGE_SCRIPT_ID } from "../vite-plugin-scripts/index.js";
import { getStylesForURL } from "./css.js";
import { getComponentMetadata } from "./metadata.js";
import { createResolve } from "./resolve.js";
import { default404Page } from "./response.js";
import { getScriptsForURL } from "./scripts.js";
class DevPipeline extends Pipeline {
  constructor(loader, logger, manifest, settings, config = settings.config) {
    const mode = "development";
    const resolve = createResolve(loader, config.root);
    const serverLike = isServerLikeOutput(config);
    const streaming = true;
    super(logger, manifest, mode, [], resolve, serverLike, streaming);
    this.loader = loader;
    this.logger = logger;
    this.manifest = manifest;
    this.settings = settings;
    this.config = config;
  }
  // renderers are loaded on every request,
  // so it needs to be mutable here unlike in other environments
  renderers = new Array();
  static create({
    loader,
    logger,
    manifest,
    settings
  }) {
    return new DevPipeline(loader, logger, manifest, settings);
  }
  async headElements(routeData) {
    const {
      config: { root },
      loader,
      mode,
      settings
    } = this;
    const filePath = new URL(`./${routeData.component}`, root);
    const { scripts } = settings.config.experimental.directRenderScript ? { scripts: /* @__PURE__ */ new Set() } : await getScriptsForURL(filePath, settings.config.root, loader);
    if (isPage(filePath, settings) && mode === "development") {
      scripts.add({
        props: { type: "module", src: "/@vite/client" },
        children: ""
      });
      if (settings.config.devToolbar.enabled && await settings.preferences.get("devToolbar.enabled")) {
        const src = await resolveIdToUrl(loader, "astro/runtime/client/dev-toolbar/entrypoint.js");
        scripts.add({ props: { type: "module", src }, children: "" });
        const additionalMetadata = {
          root: url.fileURLToPath(settings.config.root),
          version: ASTRO_VERSION,
          latestAstroVersion: settings.latestAstroVersion,
          debugInfo: await getInfoOutput({ userConfig: settings.config, print: false })
        };
        const children = `window.__astro_dev_toolbar__ = ${JSON.stringify(additionalMetadata)}`;
        scripts.add({ props: {}, children });
      }
    }
    for (const script of settings.scripts) {
      if (script.stage === "head-inline") {
        scripts.add({
          props: {},
          children: script.content
        });
      } else if (script.stage === "page" && isPage(filePath, settings)) {
        scripts.add({
          props: { type: "module", src: `/@id/${PAGE_SCRIPT_ID}` },
          children: ""
        });
      }
    }
    const links = /* @__PURE__ */ new Set();
    const { urls, styles: _styles } = await getStylesForURL(filePath, loader);
    for (const href of urls) {
      links.add({ props: { rel: "stylesheet", href }, children: "" });
    }
    const styles = /* @__PURE__ */ new Set();
    for (const { id, url: src, content } of _styles) {
      scripts.add({ props: { type: "module", src }, children: "" });
      styles.add({ props: { "data-vite-dev-id": id }, children: content });
    }
    return { scripts, styles, links };
  }
  componentMetadata(routeData) {
    const {
      config: { root },
      loader
    } = this;
    const filePath = new URL(`./${routeData.component}`, root);
    return getComponentMetadata(filePath, loader);
  }
  async preload(filePath) {
    const { loader } = this;
    if (filePath.href === new URL(DEFAULT_404_COMPONENT, this.config.root).href) {
      return { default: default404Page };
    }
    const renderers__ = this.settings.renderers.map((r) => loadRenderer(r, loader));
    const renderers_ = await Promise.all(renderers__);
    this.renderers = renderers_.filter((r) => Boolean(r));
    try {
      return await loader.import(viteID(filePath));
    } catch (error) {
      if (MarkdownError.is(error) || CSSError.is(error) || AggregateError.is(error)) {
        throw error;
      }
      throw enhanceViteSSRError({ error, filePath, loader });
    }
  }
  clearRouteCache() {
    this.routeCache.clearAll();
  }
}
export {
  DevPipeline
};
